primitives/primproc/batchprimitiveprocessor.cpp

```diff
@@ -615,71 +796,78 @@ void BatchPrimitiveProcessor::addToJoiner(ByteStream& bs)
     }
 
     idbassert(bs.length() == 0);
-    addToJoinerLock.unlock();
+}
+
+void BatchPrimitiveProcessor::doneSendingJoinerData()
+{
+	/*  to get wall-time of hash table construction
+    if (!firstCallTime.is_not_a_date_time() && !(sessionID & 0x80000000))
+    {
+        boost::posix_time::ptime now = boost::posix_time::microsec_clock::universal_time();
+        Logger logger;
+        ostringstream os;
+        os << "id " << uniqueID << ": joiner construction time = " << now-firstCallTime;
+        logger.logMessage(os.str());
+        cout << os.str() << endl;
+    }
+	*/
 }
 
 int BatchPrimitiveProcessor::endOfJoiner()
 {
     /* Wait for all joiner elements to be added */
     uint32_t i;
-
-    boost::mutex::scoped_lock scoped(addToJoinerLock);
+    size_t currentSize;
+    // it should be safe to run this without grabbing this lock
+    //boost::mutex::scoped_lock scoped(addToJoinerLock);
 
     if (endOfJoinerRan)
         return 0;
 
+    // minor hack / optimization.  The instances not inserting the table data don't
+    // need to check that the table is complete.
+    if (!firstInstance)
+    {
+        endOfJoinerRan = true;
+        pthread_mutex_unlock(&objLock);
+        return 0;
+    }
+
     if (ot == ROW_GROUP)
         for (i = 0; i < joinerCount; i++)
         {
             if (!typelessJoin[i])
             {
-                if ((tJoiners[i].get() == NULL || tJoiners[i]->size() !=
-                        tJoinerSizes[i]))
+                currentSize = 0;
```

Comment:
I might have missed the difference b/w these if-else blocks. Why do we need this if-else if the blocks are the same?